### ISSUE TITLE
False positive `redefined-outer-name` for variables bound to the exception in except clauses

### ISSUE THREAD

**Issue by MikhailRyazanov** [pylint-dev/pylint#9671]
### Bug description

```python
# pylint: disable=missing-module-docstring, missing-function-docstring
def f():
    try:
        raise ValueError('inner')
    except ValueError as e:
        print(e)

try:
    raise ValueError('outer')
except ValueError as e:
    print(e)

# print(e)  # NameError: name 'e' is not defined
```

### Command used

```shell
pylint a.py
```

### Pylint output

```python
************* Module a
a.py:5:4: W0621: Redefining name 'e' from outer scope (line 10) (redefined-outer-name)
```

### Expected behavior

Pylint must not complain about reusing variables bound to the exception in except clauses because they are in fact not defined outside the corresponding except clause (as can be seen by uncommenting the last line in the example).

### Pylint version

```shell
pylint 3.2.2
astroid 3.2.2
Python 3.12.3 (main, Apr 10 2024, 05:33:47) [GCC 13.2.0]
```

**Comment by MikhailRyazanov** [comment]
@mbyrnepr2, this attempt:
https://github.com/pylint-dev/pylint/blob/805663a8493166812322f86a52b7df4490fbf09a/pylint/checkers/variables.py#L1523-L1524
doesn't make much sense because the same behavior is observed when the order is reversed:
```python
# pylint: disable=missing-module-docstring, missing-function-docstring
try:
    raise ValueError('outer')
except ValueError as e:
    print(e)

def f():
    try:
        raise ValueError('inner')
    except ValueError as e:
        print(e)
```
```
a.py:10:4: W0621: Redefining name 'e' from outer scope (line 4) (redefined-outer-name)
```
Pylint should really understand that the variable from `except ... as ...:` doesn't exist outside the except block.
If it would help, this is what [Python documentation](https://docs.python.org/3/reference/compound_stmts.html#except-clause) says:
> When an exception has been assigned using `as target`, it is cleared at the end of the `except` clause. This is as if
> ```python
> except E as N:
>    foo
>```
> was translated to
> ```python
> except E as N:
>    try:
>        foo
>    finally:
>        del N
>```
As a side note, Pylint apparently doesn't understand explicit `del` either (I'm not saying that using it to redefine variables can be considered “good”, but if Pylint had working logic for `del`, it could be used for correcting this `except` problem).

**Comment by mbyrnepr2** [comment]
Thank you @MikhailRyazanov. That indeed makes plenty of sense!
I've taken an alternative approach and perhaps this can be a workable solution for this case.
FYI since you mentioned `del` there is some reading you can check out [here](https://github.com/pylint-dev/pylint/issues/841) on that.
